### PR TITLE
04 get article

### PR DESCRIPTION
**Pull request by OniUnderscore**

### CODE REVIEW

File: __tests__/article.test.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+describe("04 GET /api/articles/:article_id", () => {
+  describe("Functionality", () => {
+    test("Should return an article object ", () => {
+      return request(app)
+        .get("/api/articles/1")
+        .expect(200)
+        .then(({ body }) => {
+          const { article } = body;
+          expect(article.constructor).toEqual(Object);
+        });
```
Comment (nc-mitch-wyatt) [inline __tests__/article.test.js:25]: I didn't know `.constructor` was something you could access to make an assertion based on the Object class type 

I think this shows an impressive understanding of testing matchers, however it is made slightly redundant by the next test which also implicitly proves it must be an object

File: __tests__/article.test.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+              article_id: expect.any(Number),
+              article_img_url: expect.any(String),
+              author: expect.any(String),
+              body: expect.any(String),
+              created_at: expect.any(String),
+              title: expect.any(String),
+              topic: expect.any(String),
+              votes: expect.any(Number),
+            })
+          );
```
Comment (nc-mitch-wyatt) [inline __tests__/article.test.js:45]: For this test, line 31 `.expect(200)` is repeated from the previous test and can be cut ✂️ 

Would the `article_id` be _any number_ or would it be `1` if that's the article that has been requested ? 
One could go so far to ask, would any of these values be uncertain if we went to look at this object in the test data 🤔

File: __tests__/article.test.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          "https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700",
+      };
+
+      return request(app)
+        .get("/api/articles/1")
+        .expect(200)
+        .then(({ body }) => {
+          const { article } = body;
+          expect(article).toEqual(output);
+        });
```
Comment (nc-mitch-wyatt) [inline __tests__/article.test.js:68]: This test is really the only one you need, as it completely overshadows the previous two tests before it

The only thing this test isn't doing is using `toMatchObject` to ensure that additional properties won't cause it to fail if they are added later in development

File: __tests__/article.test.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+  describe("Error Handling", () => {
+    test("If ID is not a valid format (A number), return a 300 error", () => {
+      return request(app)
+        .get("/api/articles/notavalidID")
+        .expect(400)
+        .then(({ body }) => {
+          expect(body.msg).toEqual("ID is Malformed");
+        });
```
Comment (nc-mitch-wyatt) [inline __tests__/article.test.js:79]: This is a great test, check test description for typo it should be `400`

File: __tests__/article.test.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    });
+    test("If ID is valid, but has no content associated with it, should return a 404", () => {
+      return request(app)
+        .get("/api/articles/6565")
+        .expect(404)
+        .then(({ body }) => {
+          expect(body.msg).toEqual("Article not Found");
```
Comment (nc-mitch-wyatt) [inline __tests__/article.test.js:86]: Also a great error test

File: API/controllers/article-controller.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      return fetchArticle(article_id);
+    })
+    .then(({ rows }) => {
+      if (rows.length === 0) {
+        return Promise.reject({ status: 404, msg: "Article not Found" });
+      }
+      const article = rows[0];
+      res.status(200).send({ article });
+    })
+    .catch(next);
```
Comment (nc-mitch-wyatt) [inline API/controllers/article-controller.js:21]: There are a few things here:
- Firstly this is a great handling of all the logic
- On a `400` where the ID is `NaN` , if the query is made with the invalid ID then `psql` will throw an error by default, which can then be handled by your error handling middleware checking for a code of `"22P02"`. So the meta-model logic is not required to handle this error
- This leads onto another question... if the model will reject a bad query, should it also reject from there if `rows.length === 0`. In order to keep effective separation of concerns I would move this logic over to the model, as it's already dealing with one error it should probably deal with them both

**Review by nc-mitch-wyatt — COMMENTED** [review]
Lots of comments below 👇 

TLDR is, testing could be condensed & controller is doing more than it needs to do
